### PR TITLE
Add Identity Recorded tests

### LINKED ISSUE #34427: Add Recorded Tests for Identity
Add Recorded tests for azure-identity package.
Cover all scenarios where requests can be recorded and tested.

### PR DESCRIPTION
**Pull request by g2vinay**
Resolves #34427

Adds Recorded Tests for:
- Client Certificate Credential
- Client Secret Credential
- Client Assertion Credential

### CODE REVIEW

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/IdentityTestBase.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        + "YWxnIjoiUlMyNTYifQ.eyJpc3MiOiIwNmM4MmFjZC1hMDIzLTQwMGQtYjZhOC1kNjU3MDQ5NzliNGYiLCJhdWQiOiJodHRwczovL2xvZ2l"
+        + "uLm1pY3Jvc29mdG9ubGluZS5jb20vNDZkZDVhNDMtZDE5YS00YWU0LTljNjAtNjUwY2M4OTA5YjExL29hdXRoMi92Mi4wL3Rva2VuIiwic"
+        + "3ViIjoiMDZjODJhY2QtYTAyMy00MDBkLWI2YTgtZDY1NzA0OTc5YjRmIiwibmJmIjoxNzA4NjUyNjQyLCJleHAiOjE3MDg2NTMyNDIsImp"
+        + "0aSI6ImQ4YTVlZThlLWZmNzMtNDdmZC05NTg0LTFiZmI3NTc3NDc2MiJ9.EKjPiv89K7_awBtOSrguQ9BUIbO_RylvyPuH8a6u-N-6FdX3"
+        + "dG3V9fEnR7PEId8yZnQq4QAGyFirmf9vPy8XXdJ1h-Ok8PzFcU-FtN-aFddRhDBZEj37kXtqyNqEq-lw4eQvuURCQrk_e8ZsG6XR2SZsyTM"
+        + "uBfr_maQi0Uagg-yax9_ITK1OmJhfv0e93H4zBNsCucT-LFT2IyvXaULBoq04HyhFkhlvlyC8pSiM9jqTYwm64y0ipG-LHbq1jmwHdyTAXx"
+        + "OtYfPjXZTAHZD2NakTZQjf3-TC-Ol-7xuD5rtj749RApstk3YglQ5L2rf1e989nu7Jvuh8-XLkz_wrIe06RaBeZsztS-yg3ZlrfOB34glq7"
+        + "YpjRAYQXcnnHzfLLibxmMVY0bL1nLRR1PXmgBX2udxpHdm49CwaEXzO4RXPCKMwndFktkxNCv8yXUI1lhZWesVXmVns4RnGF_3HI8J3peS-"
+        + "JQ6b3ZYgekD12tJ54GxaebjXXepQz9AHyfRVPJjayT4YBb7V4Gtq1qZhNi44BFx0f-gaZdkBQhx2eaMRxFjJ9lqPTEYWHO0G2gcfH6MIL7r"
+        + "EwMfQ30ZjDYuVfiiAe7FE8L1ANxWsXvmNwYrRC7U_QCHXl7nwnflVb_1Isd-T2E-bc6z3jFbLyLlE4SYzP6468GlhlCwajIybEME";
```
Comment (mssfang) [inline sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/IdentityTestBase.java:34]: We can put those Fake credential into a file and suppress the file, https://github.com/Azure/azure-sdk-for-java/blob/main/eng/CredScanSuppression.json#L59


https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/implementation/FakeCredentialConstants.java

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/IdentityTestBase.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        if (interceptorManager.isRecordMode()) {
+            policies.add(interceptorManager.getRecordPolicy());
+        }
+
+        if (interceptorManager.isPlaybackMode()) {
+            List<TestProxyRequestMatcher> customMatchers = new ArrayList<>();
+            customMatchers.add(new BodilessMatcher());
+            customMatchers.add(new CustomMatcher().setExcludedHeaders(Collections.singletonList("X-MRC-CV")));
+            interceptorManager.addMatchers(customMatchers);
```
Comment (billwert) [inline sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/IdentityTestBase.java:66]: `addMatchers` is varargs so you don't need to make the list.

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/IdentityTestBase.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            customSanitizers.add(new TestProxySanitizer("x-client-last-telemetry", null, "REDACTED",
+                TestProxySanitizerType.HEADER));
+            customSanitizers.add(new TestProxySanitizer(null, "(client_id=)[^&]+", "$1Dummy-Id",
+                TestProxySanitizerType.BODY_REGEX));
+            customSanitizers.add(new TestProxySanitizer(null, "(client_secret=)[^&]+", "$1Dummy-Secret",
+                TestProxySanitizerType.BODY_REGEX));
+            customSanitizers.add(new TestProxySanitizer(null, "(client_assertion=)[^&]+", "$1Dummy-Secret",
+                TestProxySanitizerType.BODY_REGEX));
+
+            interceptorManager.addSanitizers(customSanitizers);
```
Comment (billwert) [inline sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/IdentityTestBase.java:55]: same - this is also varargs.

**Review by billwert — APPROVED** [review]
none of my comments are blocking, so lgtm.

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/ClientAssertionCredentialTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            .clientId(isPlaybackMode() ? "Dummy-Id" : getClientId())
+            .tenantId(isPlaybackMode() ? "Dummy-Id" : getTenantId())
+            .clientAssertion(() -> isPlaybackMode() ? INVALID_DUMMY_CLIENT_ASSERTION : getClientAssertion())
+            .pipeline(super.getHttpPipeline(httpClient))
+            .build();
+    }
+
+    @ParameterizedTest(name = DISPLAY_NAME_WITH_ARGUMENTS)
+    @MethodSource("getHttpClients")
+    @EnabledIfEnvironmentVariable(named = "AZURE_TEST_MODE", matches = "PLAYBACK")
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply sdk/identity/azure-identity/src/test/java/com/azure/identity/recorded/ClientAssertionCredentialTest.java:35]: yeah, the live setup will require config on pipeline side, so will add that for applicable credentials in a follow up PR.
This gets the end to end testing for CI pipelines unblocked.

**Pull request merged**